Design/AlignDetectors_MoreRework.md

```diff
@@ -0,0 +1,51 @@
+Motivation
+==========
+
+The ISIS powder diffraction scientists have identified a problem with their powder diffraction reduction workflow. The workflow is as follows (algorithm names in brackets):
+
+* calibration ([CrossCorrelate](https://docs.mantidproject.org/nightly/algorithms/CrossCorrelate.html), [GetDetectorOffsets](https://docs.mantidproject.org/nightly/algorithms/GetDetectorOffsets.html)). Various approaches available here as described in the Mantid [concept page](https://docs.mantidproject.org/nightly/concepts/calibration/PowderDiffractionCalibration.html). ISIS tend to use CrossCorrelate\GetDetectorOffsets which generates a "per detector" correction in the form of a peak offset. The peak offset can easily be converted into a DIFC diffractometer constant. An alternative is the PDCalibration algorithm that directly outputs DIFC as well as DIFA\TZERO
+* convert data from TOF to d spacing ([AlignDetectors](https://docs.mantidproject.org/nightly/algorithms/AlignDetectors.html)). Uses the diffractometer constants generated in the calibration step
+* focus ([DiffractionFocussing](https://docs.mantidproject.org/nightly/algorithms/DiffractionFocussing.html)). Focus the data for each bank of detectors according to grouping data in a .cal file
+* convert data back to TOF ([ConvertUnits](https://docs.mantidproject.org/nightly/algorithms/ConvertUnits-v1.html))
+* Save data to file for downstream processing ([SaveGSS](https://docs.mantidproject.org/nightly/algorithms/SaveGSS-v1.html))
+
+I believe there is a similar workflow followed at SNS using the [AlignAndFocusPowder](https://docs.mantidproject.org/nightly/algorithms/AlignAndFocusPowder-v1.html) workflow algorithm. ISIS use some python scripts to chain together the different steps but it achieves a similar process to AlignAndFocusPowder
+
+The problem is that the final ConvertUnits step to convert from d spacing to TOF doesn't use the diffractometer constants. Following the focussing the data isn't "per detector" any more but it should still be possible to use a "per bank" DIFC\DIFA\TZERO to convert the data back to TOF

```

Comment:
`AlignAndFocusPowder` uses an effective detector position to convert back to TOF (if supplied) or the center-of-mass of the pixels contributing to the focused spectrum. The effective postions are specified using `PrimaryFlightPath`, `L2`, `Polar`, and `Azimuthal` parameters of `AlignAndFocusPowder`/`AlignAndFocusPowderFromFiles`. Not specifying these values will give the default behavior (center of mass position) which is essentially nonsense for a 2nd generation (and beyound) powder instrument.

Reply:
> Lots of comments on this one, maybe I've been here too long. I think most of this can be attributed to not fully understanding all of the parameters in `AlignAndFocusPowder`.
> 
> 1. Using the properties `PrimaryFlightPath`, `L2`, `Polar`, and `Azimuthal` sets the focus position of all of the output spectra via [EditInstrumentGeometry](https://docs.mantidproject.org/nightly/algorithms/EditInstrumentGeometry-v1.html). The resulting workspace will work exactly as expected in [ConvertUnits](https://docs.mantidproject.org/nightly/algorithms/ConvertUnits-v1.html). The `SpectrumIDs` parameter is the other confusing one that may be easier to understand by reading the code.
> 2. "Offsets" are very useful for seeing how much parameters are changed, but are relative to a previous instrument geometry. We had written some code once that allowed for calculating offsets between any two calibrations `(new - old)/old` which was fairly useful for seeing how the instrument evolved over time. The change of `AlignAndFocusPowder` to use DIFC was to break this dependency and allow for new methods of calibration which can start from a previous calibration. [PDCalibration](https://docs.mantidproject.org/nightly/algorithms/PDCalibration-v1.html) was designed with this new thinking.
> 3. `DIFC ~ L2 sin(theta)` cannot be converted back to a detector position. [AlignComponents](https://docs.mantidproject.org/nightly/algorithms/AlignComponents-v1.html) will move instrument components around until the DIFC from the detector position is as close as possible to those from calibration.
> 4. Having DIFC available in the instrument view is an interesting extra bit of information.
> 5. Using `ConvertUnits` at the end of `AlignAndFocusPowder` will use the same "wrong geometry" to go back to d-spacing. Then one can use `EditInstrumentInfo` and convert back to TOF. Again, the extra parameters on `AlignAndFocusPowder` do this already.

Thanks a lot for taking a look at this.

I've had another look at the AlignAndFocusPowder algorithm and I see the EditInstrumentGeometry step that I'd missed before and I can see that would solve the problem if you have a suitable effective position for each focussed group. The ISIS powder calibration workflow doesn't involve producing an effective position however. We're just using the offsets approach to represent the calibrated detector positions vs the IDF as a peak offset. So the problem I'm trying to fix may be ISIS-specific. I guess we could change the workflow to be based around effective detector positions but in my discussions with the scientists here I had the impression this could be fixed without a change in the calibration approach.

ISIS don't use AlignAndFocusPowder but we do run a similar chain of algorithms. I think the scenario that the ISIS instruments use is closest to running AlignAndFocusPowder with the CalibrationWorkspace parameter supplied (and no effective positions supplied). In this case the final conversion from d spacing to TOF looks to me as though it ignores the DIFCs in the CalibrationWorkspace and just uses a DIFC based on the nominal detector positions from the IDF. It seems like ConvertUnits should use the DIFCs from the CalibrationWorkspace during the final conversion to TOF in this case?

I don't think the changes I'm proposing would get in the way of a workflow based on effective positions. I would blank out any DIFCs in the instrument parameter map whenever any of the detector positions are changed eg if EditInstrumentGeometry is run. So ConvertUnits would do what it's always done in your workflow even if I make these changes.